documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,329 @@
+        /// </returns> 
+        public static GenerateCoreDump(int processId, string dumpName, DumpType dumpType, bool logDumpGeneration)
+
+        /// <summary>
+        /// Attach a profiler.
+        /// </summary>
+        /// <param name="processId">Target process' ID</param>
+        /// <param name="attachTimeout">Timeout in seconds</param>
+        /// <param name="attachTimeout">
+        public static int AttachProfiler(int processId, uint attachTimeout, Guid profilerGuid, string profilerPath, byte[] additionalData);
+
+    }
+}
+```
+
+
+### Exceptions that can be thrown 
+
+```cs
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    // Generic wrapper for exceptions thrown by this library
+    public class DiagnosticsClientExcpetion : Exception {} 
+
+    // When a certian command is not supported by either the library or the target process' runtime
+    public class UnsupportedProtocolException : DiagnosticsClientException {} 
+
+    // When the runtime is no longer availble for attaching.
+    public class ServerNotAvailableException : DiagnosticsClientException {} 
+
+    // When the runtime responded with an error
+    public class ServerThrownException : DiagnosticsClientException {} 
+}
+```
+
+## Micorosft.Diagnostics.Client.EventPipe namespace
+
+### EventPipeProvider
+A class that describes an EventPipe provider.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class EventPipeProvider
+    {
+        public EventPipeProvider(
+            string name,
+            long keywords = ulong.MaxValue,
+            EventLevel eventLevel = EventLevel.Verbose,
+            string filterData = null)
+
+        public long Keywords { get; }
+
+        public EventLevel EventLevel { get; }
+
+        public string Name { get; }
+
+        public IEnumerable<KeyValuePair<string, string>> FilterData { get; }
+
+        public override string ToString();
+        
+        public override bool Equals(object obj);
+
+        public override int GetHashCode();
+
+        public static bool operator ==(Provider left, Provider right);
+
+        public static bool operator !=(Provider left, Provider right);
+    }
+}
+```
+
+### EventPipeSession
+This is a class to represent an EventPipeSession. It is meant to be immutable and acts as a handle to each session that has been started. 
+
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class EventPipeSession
+    {
+        public EventPipeSession(
+            int processId,
+            long sessionId,
+            IEnumerable<EventPipeProvider> providers,
+            int circularBufferMB,
+            bool rundownRequested,
+            Stream stream
+        )
+        public int ProcessId { get; }
+        public int SessionId { get; } 
+        public ReadOnlyList<EventPipeProvider> Providers { get ; }
+        public int CircularBufferMB { get; }
+        public bool RundownRequested { get; }
+        public Stream Stream { get; };
```

Comment:
Do we want to put a `Stop` method on here that calls through to the `client`?

Reply:
We could. In that case I'd rather remove the `DiagnosticsClient.StopTracing` since that would cause confusion.